File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/android/src/main/kotlin/com/idosmart/pigeon_implement/AlexaImpl.kt
package com.idosmart.pigeon_implement

import com.idosmart.pigeongen.api_alexa.*
import com.idosmart.protocol_channel.plugin
import com.idosmart.protocol_sdk.*
import com.idosmart.enums.*
import com.idosmart.protocol_channel.innerRunOnMainThread

class AlexaLoginCancellable(override var isCancelled: Boolean = false) : IDOCancellable {

    private fun alexa(): Alexa? {
        return plugin.alexa()
    }

    override fun cancel() {
        isCancelled = true
        alexa()?.stopLogin {}
    }

}

internal class AlexaImpl: AlexaInterface, IDOAlexaAuthDelegate {

    private var _handleBlock: ((String,String) -> Unit)? = null
    private var _loginStateBlock: ((IDOAlexaLoginState) -> Unit)? = null

    private fun alexa(): Alexa? {
        return plugin.alexa()
    }

    private var _isLogin = false

    override val isLogin: Boolean
        get() = _isLogin

    override fun onLoginStateChanged(handle: (IDOAlexaLoginState) -> Unit) {
        _loginStateBlock = handle
    }

    override fun onNetworkChanged(hasNetwork: Boolean) {
        innerRunOnMainThread {
            alexa()?.onNetworkChanged(hasNetwork) {}
        }
    }

    override fun setupAlexa(delegate: IDOAlexaDelegate, clientId: String) {
        AlexaDelegateImpl.instance().delegate = delegate
        AlexaAuthDelegateImpl.instance().delegate = this
        innerRunOnMainThread {
            alexa()?.registerAlexa(clientId) {}
        }
    }

    override fun authorizeRequest(
        productId: String,
        handle: (verificationUri: String, pairCode: String) -> Unit,
        completion: (rs: IDOAlexaAuthorizeResult) -> Unit
    ): IDOCancellable {
        _handleBlock = handle
        innerRunOnMainThread {
            alexa()?.authorizeRequest(productId) {
                completion(
                    IDOAlexaAuthorizeResult.ofRaw(it.toInt()) ?: IDOAlexaAuthorizeResult.SUCCESSFUL
                )
            }
        }
        return AlexaLoginCancellable()
    }

    override fun logout() {
        innerRunOnMainThread {
            alexa()?.logout {}
        }
    }

    override fun callbackPairCode(userCode: String, verificationUri: String) {
        _handleBlock?.let { it(verificationUri,userCode) }
    }

    override fun loginStateChanged(state: IDOAlexaLoginState) {
        val loginState = IDOAlexaLoginState.ofRaw(state.raw) ?: IDOAlexaLoginState.LOGGING
        when (loginState) {
            IDOAlexaLoginState.LOGGING -> _isLogin = false
            IDOAlexaLoginState.LOGINED -> _isLogin = true
            IDOAlexaLoginState.LOGOUT -> _isLogin = false
        }
        _loginStateBlock?.let { it(loginState) }
    }


}

File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/ios/Classes/protocol_sdk/impl/AlexaImpl.swift
//
//  AlexaImpl.swift
//  protocol_channel
//
//  Created by hc on 2023/7/18.
//

import Foundation

private var _flagSetupAlexa = false
private var _alexa: Alexa? {
    return SwiftProtocolChannelPlugin.shared.alexa
}

class AlexaImpl: NSObject, IDOAlexaInterface {
    
    private var _handleBlock: ((String, String) -> Void)?
    private var _loginStateBlock: ((IDOAlexaLoginState) -> Void)?
    public private(set) var isLogin: Bool = false
    
    func onNetworkChanged(hasNetwork: Bool) {
        _runOnMainThread {
            _alexa?.onNetworkChanged(hasNetwork: hasNetwork, completion: { })
        }
    }
    
    func onLoginStateChanged(handle : @escaping (IDOAlexaLoginState) -> Void) {
        _loginStateBlock = handle
    }
    
    func setupAlexa(delegate: IDOAlexaDelegate, clientId: String) {
        AlexaDelegateImpl.shared.delegate = delegate
        AlexaAuthDelegateImpl.shared.delegate = self
        if (_flagSetupAlexa) {
            _runOnMainThread {
                _logNative("重复调用setupAlexa(...)")
            }
            return
        }
        _flagSetupAlexa = true
        _runOnMainThread {
            _alexa?.registerAlexa(clientId: clientId) {}
        }
    }
    
    func authorizeRequest(productId: String,
                          handle: @escaping (_ verificationUri: String, _ pairCode: String) -> Void,
                          completion: @escaping (_ rs: IDOAlexaAuthorizeResult) -> Void) -> IDOCancellable {
        _handleBlock = handle
        _runOnMainThread {
            _alexa?.authorizeRequest(productId: productId, completion: { rs in
                completion(IDOAlexaAuthorizeResult.init(rawValue: rs.int)!)
            })
        }
        return AlexaLoginCancellable()
    }

    func logout() {
        _runOnMainThread {
            _alexa?.logout {}
        }
    }
}

// MARK: -

class AlexaLoginCancellable: IDOCancellable {
    var isCancelled = false
    func cancel() {
        isCancelled = true
        _runOnMainThread {
            _alexa?.stopLogin {}
        }
    }
}

extension AlexaImpl: IDOAlexaAuthDelegate {
    func callbackPairCode(userCode: String, verificationUri: String) {
        _handleBlock?(verificationUri, userCode)
        _handleBlock = nil
    }
    
    func loginStateChanged(state: IDOAlexaLoginState) {
        switch(state) {
        case .logging:
            isLogin = false
        case .logined:
            isLogin = true
        case .logout:
            isLogin = false
        }
        _loginStateBlock?(state)
    }
    
}

/// alexa授权登录结果
@objc
public enum IDOAlexaAuthorizeResult: Int {
    /// 成功
    case successful = 0
    
    /// 失败
    case failure = 1
    
    /// 超时
    case timeout = 3
}
